sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3676,3 +3678,63 @@ object UpdateOuterReferences extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Resolves different children of Union to a common set of columns. Note that this must be
+ * run before `TypeCoercion`, because `TypeCoercion` should be run on correctly resolved
```

Comment:
It's fragile to rely on rule order. How about we skip the type coercion rule if the union is by name and the name match is not done yet?

Reply:
Ok.